README.md
p-net Profinet device stack
===========================

Web resources
-------------

* Source repository: [https://github.com/rtlabs-com/p-net](https://github.com/rtlabs-com/p-net)
* Documentation: [https://rt-labs.com/docs/p-net](https://rt-labs.com/docs/p-net)
* Continuous integration: [https://github.com/rtlabs-com/p-net/actions](https://github.com/rtlabs-com/p-net/actions)
* RT-Labs (stack integration, certification services and training): [https://rt-labs.com](https://rt-labs.com)


[![Build Status](https://github.com/rtlabs-com/p-net/workflows/Build/badge.svg?branch=master)](https://github.com/rtlabs-com/p-net/actions?workflow=Build)
[![CodeQL](https://github.com/rtlabs-com/p-net/workflows/CodeQL/badge.svg?branch=master)](https://github.com/rtlabs-com/p-net/actions?workflow=CodeQL)

Converting examples for your own usage
-------------------------------------
Convert p-net to use it with your own datatypes and GSDML file.

1. Copy the GSML file that is found in the `p-net/samples/pn_dev/GSDML-V2.4-RT-Labs-P-Net-Sample-App-20220324.xml`
    Note: I tried to create my own profinet device with TwinCAT or with CODESYS and export a GSDML file.
    But I never got this to work and I don´t know why.
2. Manually adjust the GSDML file as needed, for example:
   - the company name, name of the device etc.
   - size of IO, datatypes of IO and number of modules
   - also make sure all submodules have a unique ID (not sure if needed).
1. Duplicate the `samples/pn_simple_example` folder and content and rename the `pn_simple_example` folder, for example to `pn_my_app`
2. Then open the `CMakeLists.txt` in your `pn_my_app` folder and rename all `pn_simple_example` to `pn_my_app`
3. Open `app_gsdml.h` and add the sizes and (sub)module IDs of the modules you added
4. In `app_gsdml.c`:
   1. Add `static const app_gsdml_module_t`  for each module. Check that all members (`.id`, `.name`, etc.) are correct
   2. Add `static const app_gsdml_submodule_t`  for each submodule and check that all the members are correct
   3. Add the added modules and sub modules to the supported lists: `app_gsdml_modules[]` and `app_gsdml_submodules[]`
5. Open `app_data.c`:
   1. add a static struct where in/output data can be stored into.
      - `static uint8_t inputdata[APP_GSDML_INPUT_DATA_DIGITAL_SIZE] = {0};`
      - Floats are saved as unsigned integers, because of endian conversion. Profinet data has [network endianess](https://en.wikipedia.org/wiki/Endianness#Networking) (Big endian), whereas C uses little endian ?
  1. add a `typedef struct` to store I/O data

Endianess = The attribute of a system that indicates whether integers are represented with the most significant byte stored at the lowest address (big endian) or at the highest address (little endian)

p-net
-----
Profinet device stack implementation. Key features:
* Profinet v2.4
  * Conformance Class A and B
  * Real Time Class 1
  * Multiple Ethernet ports
* Easy to use
  * Extensive documentation and instructions on how to get started.
  * Build and run sample application on Raspberry Pi in 30 minutes.
* Portable
  * Written in C.
  * Linux, RTOS or bare metal.
  * Sources for supported port layers provided.

The RT-Labs Profinet stack p-net is used for Profinet device
implementations. It is easy to use and provides a small footprint. It
is especially well suited for embedded systems where resources are
limited and efficiency is crucial.
The stack is supplied with full sources including porting
layers and a sample application.

Also C++ (any version) is supported for application development.

The main requirement on the platform
is that it can send and receive raw Ethernet Layer 2 frames.

Features:
 * Multiple Ethernet ports (for Linux only, so far)
 * TCP/IP
 * LLDP
 * SNMP
 * RT (real-time)
 * Address resolution
 * Parameterization
 * Process IO data exchange
 * Alarm handling
 * Configurable number of modules and sub-modules
 * Bare-metal or OS
 * Porting layer provided
 * Supports I&M0 - I&M4. The I&M data is supported for the device, but not for
   individual modules.

Limitations or not yet implemented:

* This is a device stack, which means that the IO-controller/master/PLC side is
  not supported.
* No media redundancy (No MRP support)
* Legacy startup mode is not fully implemented
* No support for RT_CLASS_UDP
* No support for DHCP
* No fast start-up
* No MC multicast device-to-device
* No support of shared device (connection to multiple controllers)
* Supports only full connections, not the limited "DeviceAccess" connection type.
* No iPar (parameter server) support
* No support for time synchronization
* No UDP frames at alarm (just the default alarm mechanism is implemented)
* No ProfiDrive or ProfiSafe profiles.

This software is dual-licensed, with GPL version 3 and a commercial license.
If you intend to use this stack in a commercial product, you likely need to
buy a license. See LICENSE.md for more details.


Requirements
------------
The platform must be able to send and receive raw Ethernet Layer 2 frames,
and the Ethernet driver must be able to handle full size frames. It
should also avoid copying data, for performance reasons.

* cmake 3.14 or later

For Linux:

* gcc 4.6 or later
* See the "Real-time properties of Linux" page in the documentation on how to
  improve Linux timing

For rt-kernel:

* Workbench 2020.1 or later

An example of microcontroller we have been using is the Infineon XMC4800,
which has an ARM Cortex-M4 running at 144 MHz, with 2 MB Flash and 352 kB RAM.
It runs rt-kernel, and we have tested it with 9 Profinet slots each
having 8 digital inputs and 8 digital outputs (one bit each). The values are
sent and received each millisecond (PLC watchdog setting 3 ms).


Getting started
---------------
See the tutorial in the documentation: [https://rt-labs.com/docs/p-net/tutorial.html](https://rt-labs.com/docs/p-net/tutorial.html)

Note that you need to include submodules when cloning:

```
git clone --recurse-submodules https://github.com/rtlabs-com/p-net.git
```

Dependencies
------------
Some of the platform-dependent parts are located in the OSAL repository and the
cmake-tools repository.

* [https://github.com/rtlabs-com/osal](https://github.com/rtlabs-com/osal)
* [https://github.com/rtlabs-com/cmake-tools](https://github.com/rtlabs-com/cmake-tools)

Those are downloaded automatically during install.

The p-net stack contains no third party components. Its external dependencies are:

* C-library
* An operating system (if used)
* For conformance class B you need an SNMP implementation. On Linux is
   net-snmp (BSD License) used [http://www.net-snmp.org](http://www.net-snmp.org)

Tools used for building, testing and documentation (not shipped in the resulting binaries):

* cmake (BSD 3-clause License)  [https://cmake.org](https://cmake.org)
* gtest (BSD-3-Clause License) [https://github.com/google/googletest](https://github.com/google/googletest)
* Sphinx (BSD license) [https://www.sphinx-doc.org](https://www.sphinx-doc.org)
* Doxygen (GPL v2) [https://www.doxygen.nl](https://www.doxygen.nl/index.html)
* clang-format (Apache License 2.0) [https://clang.llvm.org](https://clang.llvm.org/docs/ClangFormat.html)


Contributions
--------------
Contributions are welcome. If you want to contribute you will need to
sign a Contributor License Agreement and send it to us either by
e-mail or by physical mail. More information is available
on [https://rt-labs.com/contribution](https://rt-labs.com/contribution).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),

## [Unreleased]

## 2021-03-31
**Changed**
- Update GSDML file to conformance class B

## 2021-03-24
**Added**
- Possibility to configure number of ports, without recompiling.

## 2021-03-03
**Changed**
- Set thread priority and stack size via configuration

## 2021-02-19
**Changed**
- Split options.h into two files.
- Simplify configuration

## 2020-02-03
**Added**
 - Improved support for multiple ports

## 2021-01-29
**Changed**
 - Move remaining Linux sample app files to src/ports/linux

## 2020-01-11
**Added**
 - Support for multiple ports

## 2021-01-05
**Added**
 - Additional API functions for low-level diagnostic handling

## 2020-12-22
**Changed**
 - Remove link status from configuration

## 2020-12-15
**Changed**
 - Use Profinet version 2.4 for the GSDML file
 - Changed default installation path for sample_app on Raspberry Pi

## 2020-12-11
**Added**
 - Runtime data for multiple ports
 - Functionality to search for scripts in multiple locations for Linux

## 2020-12-07
**Changed**
 - Rename pnet_set_state() to pnet_set_primary_state()

## 2020-12-01
**Added**
- Changed informative callbacks to be void-functions.

## 2020-11-30
**Added**
- SNMP functionality for Linux

## 2020-11-20
**Added**
- Additional log level FATAL

## 2020-11-19
**Added**
- SNMP functionality for rt-kernel

## 2020-11-16
**Added**
- Support for multiple ports in configuration

## 2020-11-10
**Changed**
- Remove LLDP TTL from configuration
- Remove LLDP chassis ID from configuration

## 2020-11-03
**Changed**
- Improve API for adding diagnosis
- Use rtlabs-com/osal and rtlabs-com/cmake-tools as submodules

## 2020-10-28
**Changed**
- Rename compile time constants for number of slots and subslots

## 2020-09-19
**Changed**
- Modified public API for sending alarm ACK.

## 2020-08-24
**Changed**
- Renamed configuration members related to I&M data

## 2020-08-18
**Added**
- Add min_device_interval to the configuration
- Use cmake to set configurable build options

## 2020-06-15
**Added**
- New user callback for LED state change.

## 2020-06-11
**Added**
- New API function for application to perform a factory reset

## 2020-04-09
**Added**
- New user callback to indicate reset requests from the IO-controller.

## 2020-04-06
**Changed**
- Changed the user API to use a handle to the p-net stack, and to allow user
  arguments to callbacks.

## 2020-03-27
**Changed**
- Read MAC address also from Linux hardware.


src/pnal.h
/*********************************************************************
 *        _       _         _
 *  _ __ | |_  _ | |  __ _ | |__   ___
 * | '__|| __|(_)| | / _` || '_ \ / __|
 * | |   | |_  _ | || (_| || |_) |\__ \
 * |_|    \__|(_)|_| \__,_||_.__/ |___/
 *
 * www.rt-labs.com
 * Copyright 2018 rt-labs AB, Sweden.
 *
 * This software is dual-licensed under GPLv3 and a commercial
 * license. See the file LICENSE.md distributed with this software for
 * full license information.
 ********************************************************************/

/**
 * @file
 * @brief Definitions for profinet abstraction layer.
 *
 * The Ethernet-related functions should not use \a loc_port_num, but
 * the interface name instead. The lookup from \a loc_port_num to
 * interface name must happen somewhere else in the stack.
 */

#ifndef PNAL_H
#define PNAL_H

#ifdef __cplusplus
extern "C" {
#endif

#include <stddef.h>
#include <stdint.h>
#include <stdbool.h>

#include "options.h"
#include "pnet_api.h"
#include "pnal_sys.h"

#define PNAL_MAKEU32(a, b, c, d)                                               \
   (((uint32_t)((a)&0xff) << 24) | ((uint32_t)((b)&0xff) << 16) |              \
    ((uint32_t)((c)&0xff) << 8) | (uint32_t)((d)&0xff))

#define PNAL_MAKEU16(a, b) (((uint16_t)((a)&0xff) << 8) | (uint16_t)((b)&0xff))

#define PNAL_INET_ADDRSTR_SIZE 16 /** Incl termination */
#define PNAL_ETH_ADDRSTR_SIZE  18 /** Incl termination */
#ifndef PNAL_HOSTNAME_MAX_SIZE
#define PNAL_HOSTNAME_MAX_SIZE 64 /** Incl termination. Value from Linux. */
#endif

/**
 * Ethernet autonegotiation capabilities (not exhaustive)
 *
 * See IEEE 802.3 (Ethernet) ch. 30.6 "Management for link Auto-Negotiation".
 *
 * See IETF RFC 4836 "Definitions of Managed Objects for
 * IEEE 802.3 Medium Attachment Units (MAUs)", object
 * "IANAifMauAutoNegCapBits"
 */
#define PNAL_ETH_AUTONEG_CAP_1000BaseT_FULL_DUPLEX BIT (0)
#define PNAL_ETH_AUTONEG_CAP_1000BaseT_HALF_DUPLEX BIT (1)
#define PNAL_ETH_AUTONEG_CAP_1000BaseX_FULL_DUPLEX BIT (2)
#define PNAL_ETH_AUTONEG_CAP_1000BaseX_HALF_DUPLEX BIT (3)
#define PNAL_ETH_AUTONEG_CAP_100BaseTX_FULL_DUPLEX BIT (10)
#define PNAL_ETH_AUTONEG_CAP_100BaseTX_HALF_DUPLEX BIT (11)
#define PNAL_ETH_AUTONEG_CAP_10BaseT_FULL_DUPLEX   BIT (13)
#define PNAL_ETH_AUTONEG_CAP_10BaseT_HALF_DUPLEX   BIT (14)
#define PNAL_ETH_AUTONEG_CAP_UNKNOWN               BIT (15)

/**
 * Ethernet MAU type (not exhaustive).
 *
 * See IEEE 802.3 (Ethernet) ch. 30.5 "Layer management for medium
 * attachment units (MAUs)".
 *
 * See IETF RFC 4836 "Definitions of Managed Objects for
 * IEEE 802.3 Medium Attachment Units (MAUs)", objects
 * "IANAifMauTypeListBits" and "dot3MauType".
 *
 * See Profinet "5.2.12.3.12 Coding of the field MAUType"
 */
typedef enum pnal_eth_mau
{
   PNAL_ETH_MAU_UNKNOWN = 0x00, /* When link is down */
   PNAL_ETH_MAU_RADIO = 0x00,   /* When link is up */
   PNAL_ETH_MAU_COPPER_10BaseT = 0x05,
   PNAL_ETH_MAU_COPPER_100BaseTX_HALF_DUPLEX = 0x0F,
   PNAL_ETH_MAU_COPPER_100BaseTX_FULL_DUPLEX = 0x10,
   PNAL_ETH_MAU_COPPER_1000BaseT_HALF_DUPLEX = 0x1D,
   PNAL_ETH_MAU_COPPER_1000BaseT_FULL_DUPLEX = 0x1E,
   PNAL_ETH_MAU_FIBER_100BaseFX_HALF_DUPLEX = 0x11,
   PNAL_ETH_MAU_FIBER_100BaseFX_FULL_DUPLEX = 0x12,
   PNAL_ETH_MAU_FIBER_1000BaseX_HALF_DUPLEX = 0x15,
   PNAL_ETH_MAU_FIBER_1000BaseX_FULL_DUPLEX = 0x16,
} pnal_eth_mau_t;

/* See Profinet 2.4 section 5.2.28 and annex W */
typedef struct pnal_port_stats
{
   uint32_t if_in_octets;
   uint32_t if_out_octets;
   uint32_t if_in_discards;
   uint32_t if_out_discards;
   uint32_t if_in_errors;
   uint32_t if_out_errors;
} pnal_port_stats_t;

/** Set an IP address given by the four byte-parts */
#define PNAL_IP4_ADDR_TO_U32(ipaddr, a, b, c, d)                               \
   ipaddr = PNAL_MAKEU32 (a, b, c, d)

typedef enum pnal_ethertype
{
   PNAL_ETHTYPE_IP = 0x0800U,
   PNAL_ETHTYPE_ARP = 0x0806U,
   PNAL_ETHTYPE_VLAN = 0x8100U,
   PNAL_ETHTYPE_PROFINET = 0x8892U,
   PNAL_ETHTYPE_ETHERCAT = 0x88A4U,
   PNAL_ETHTYPE_LLDP = 0x88CCU,
   PNAL_ETHTYPE_ALL = 0xFFFFU,
} pnal_ethertype_t;

/* 255.255.255.255 */
#ifndef PNAL_IPADDR_NONE
#define PNAL_IPADDR_NONE ((uint32_t)0xffffffffUL)
#endif
/* 127.0.0.1 */
#ifndef PNAL_IPADDR_LOOPBACK
#define PNAL_IPADDR_LOOPBACK ((uint32_t)0x7f000001UL)
#endif
/* 0.0.0.0 */
#ifndef PNAL_IPADDR_ANY
#define PNAL_IPADDR_ANY ((uint32_t)0x00000000UL)
#endif
#define PNAL_IPADDR_INVALID PNAL_IPADDR_ANY
/* 255.255.255.255 */
#ifndef PNAL_IPADDR_BROADCAST
#define PNAL_IPADDR_BROADCAST ((uint32_t)0xffffffffUL)
#endif

typedef uint32_t pnal_ipaddr_t;
typedef uint16_t pnal_ipport_t;

/**
 * The Ethernet MAC address.
 *
 * TODO: Remove this and use pnet_ethaddr_t instead
 */
typedef struct pnal_ethaddr
{
   uint8_t addr[6];
} pnal_ethaddr_t;

/**
 * Ethernet link status.
 *
 * See also type pf_lldp_link_status_t.
 */
typedef struct pnal_eth_status_t
{
   /* Is autonegotiation supported on this port? */
   bool is_autonegotiation_supported;

   /* Is autonegotiation supported on this port? */
   bool is_autonegotiation_enabled;

   /* Capabilities advertised to link partner during autonegotiation.
    *
    * See macros PNAL_ETH_AUTONEG_CAP_xxx.
    */
   uint16_t autonegotiation_advertised_capabilities;

   /* Operational MAU type.
    *
    * This specifies both the physical medium as well as
    * speed and duplex setting used for the link.
    */
   pnal_eth_mau_t operational_mau_type;

   /* Corresponds to linux "RUNNING" flag.
    * Up and cable connected
    */
   bool running;
} pnal_eth_status_t;

/**
 * Get system uptime from the SNMP implementation.
 *
 * This is the sysUpTime, as used by SNMP:
 * "The time (in hundredths of a second) since the network
 *  management portion of the system was last re-initialized."
 * - IETF RFC 3418 (SNMP MIB-2).
 *
 * Typically used to store the SNMP-timestamp for incoming LLDP-packets.
 * This timestamp is later reported back to the SNMP implementation, and
 * remote SNMP managers might use it for filtering.
 *
 * The returned value should be the one used by the SNMP implementation.
 * Ask the SNMP implementation for the value, or make sure that this
 * function and the SNMP implementation use the same mechanism to get the
 * value.
 *
 * Starts at 0, with wrap-around after ~497 days.
 *
 * @return System uptime, in units of 10 milliseconds.
 */
uint32_t pnal_get_system_uptime_10ms (void);

/**
 * Load a binary file.
 *
 * Can load the data into two buffers.
 *
 * @param fullpath         In:    Full path to the file
 * @param object_1         Out:   Data to load, or NULL. Mandatory if size_1 > 0
 * @param size_1           In:    Size of object_1.
 * @param object_2         Out:   Data to load, or NULL. Mandatory if size_2 > 0
 * @param size_2           In:    Size of object_2.
 * @return  0  if the operation succeeded.
 *          -1 if not found or an error occurred.
 */
int pnal_load_file (
   const char * fullpath,
   void * object_1,
   size_t size_1,
   void * object_2,
   size_t size_2);

/**
 * Save a binary file.
 *
 * Can handle two output buffers.
 *
 * @param fullpath         In:    Full path to the file
 * @param object_1         In:    Data to save, or NULL. Mandatory if size_1 > 0
 * @param size_1           In:    Size of object_1.
 * @param object_2         In:    Data to save, or NULL. Mandatory if size_2 > 0
 * @param size_2           In:    Size of object_2.
 * @return  0  if the operation succeeded.
 *          -1 if an error occurred.
 */
int pnal_save_file (
   const char * fullpath,
   const void * object_1,
   size_t size_1,
   const void * object_2,
   size_t size_2);

/**
 * Clear a binary file.
 *
 * @param fullpath         In:    Full path to the file
 */
void pnal_clear_file (const char * fullpath);

/*
 */

/**
 * Allocate a buffer
 *
 * The resulting \a pnal_buf_t  buffer is defined for
 * each operating system, and should at least contain:
 *     void * payload;
 *     uint16_t len;
 *
 * @param length           In:    Length, in bytes
 * @return a pnal_buf_t, or NULL at failure
 */
pnal_buf_t * pnal_buf_alloc (uint16_t length);

/**
 * Free a buffer
 *
 * @param p           In:    Buffer to free
 */
void pnal_buf_free (pnal_buf_t * p);

/** Not yet used */
uint8_t pnal_buf_header (pnal_buf_t * p, int16_t header_size_increment);

/**
 * Network interface handle, forward declaration.
 */
typedef struct pnal_eth_handle pnal_eth_handle_t;

/**
 * The prototype of raw Ethernet reception call-back functions.
 *
 * @param eth_handle       InOut: Network interface handle
 * @param arg              InOut: User-defined (may be NULL).
 * @param p_buf            InOut: The incoming Ethernet frame
 *
 * @return  0  If the frame was NOT handled by this function.
 *          1  If the frame was handled and the buffer freed.
 */
typedef int (pnal_eth_callback_t) (
   pnal_eth_handle_t * eth_handle,
   void * arg,
   pnal_buf_t * p_buf);

/**
 * Get status of Ethernet link on specified port
 *
 * @param interface_name   In:    Ethernet interface name, for example eth0
 * @param status           Out:   Returned link status (autoneg etc)
 * @return  0 if the operation succeeded.
 *         -1 if an error occurred.
 */
int pnal_eth_get_status (
   const char * interface_name,
   pnal_eth_status_t * status);

/**
 * Get network interface index
 *
 * The interface index (ifIndex) is the (row) index for the network interface
 * in the table ifTable, which is part of the SNMP MIB-II data structure.
 * See RFC 2863 "The Interfaces Group MIB".
 *
 * @param interface_name   In:    Ethernet interface name, for example eth0
 *
 * @return  The interface index, or 0 if not available.
 */
int pnal_get_interface_index (const char * interface_name);

/**
 * Get network interface (port) statistics
 *
 * In Profinet naming a physical Ethernet interface is a "port".
 * This corresponds to for example a Linux Ethernet interface.
 *
 * @param interface_name   In:    Ethernet interface name for example eth0
 * @param port_stats       Out:   Returned statistics
 *
 * @return  0  if the operation succeeded.
 *          -1 if an error occurred.
 */
int pnal_get_port_statistics (
   const char * interface_name,
   pnal_port_stats_t * port_stats);

/**
 * Send raw Ethernet data
 *
 * @param handle           In:    Ethernet handle
 * @param buf              In:    Buffer with data to be sent
 * @return  The number of bytes sent, or -1 if an error occurred.
 */
int pnal_eth_send (pnal_eth_handle_t * handle, pnal_buf_t * buf);

/**
 * Initialize receiving of raw Ethernet frames on one interface (in separate
 * thread)
 *
 * @param if_name          In:    Ethernet interface name
 * @param receive_type     In:    Ethernet frame types that shall be received
 *                                by the network interface / port.
 * @param pnal_cfg         In:    Operating system dependent configuration
 * @param callback         In:    Callback for received raw Ethernet frames
 * @param arg              InOut: User argument passed to the callback
 *
 * @return  the Ethernet handle, or NULL if an error occurred.
 */
pnal_eth_handle_t * pnal_eth_init (
   const char * if_name,
   pnal_ethertype_t receive_type,
   const pnal_cfg_t * pnal_cfg,
   pnal_eth_callback_t * callback,
   void * arg);

/**
 * Open an UDP socket
 *
 * @param addr             In:    IP address to listen to. Typically used with
 *                                PNAL_IPADDR_ANY.
 * @param port             In:    UDP port to listen to.
 * @return Socket ID, or -1 if an error occurred.
 */
int pnal_udp_open (pnal_ipaddr_t addr, pnal_ipport_t port);

/**
 * Send UDP data
 *
 * @param id               In:    Socket ID
 * @param dst_addr         In:    Destination IP address
 * @param dst_port         In:    Destination UDP port
 * @param data             In:    Data to be sent
 * @param size             In:    Size of data
 * @return  The number of bytes sent, or -1 if an error occurred.
 */
int pnal_udp_sendto (
   uint32_t id,
   pnal_ipaddr_t dst_addr,
   pnal_ipport_t dst_port,
   const uint8_t * data,
   int size);

/**
 * Receive UDP data.
 *
 * This is a nonblocking function, and it
 * returns 0 immediately if no data is available.
 *
 * @param id               In:    Socket ID
 * @param src_addr         Out:   Source IP address
 * @param src_port         Out:   Source UDP port
 * @param data             Out:   Received data
 * @param size             In:    Size of buffer for received data
 * @return  The number of bytes received, or -1 if an error occurred.
 */
int pnal_udp_recvfrom (
   uint32_t id,
   pnal_ipaddr_t * src_addr,
   pnal_ipport_t * src_port,
   uint8_t * data,
   int size);

/**
 * Close an UDP socket
 *
 * @param id               In:    Socket ID
 */
void pnal_udp_close (uint32_t id);

/**
 * Configure SNMP server.
 *
 * This function configures a platform-specific SNMP server as to
 * enable a connected SNMP client to read variables from the p-net stack,
 * as well as to write some variables to it.
 *
 * @param net              InOut: The p-net stack instance
 * @param pnal_cfg         In:    Operating system dependent configuration
 * @return  0 if the operation succeeded.
 *         -1 if an error occurred.
 */
int pnal_snmp_init (pnet_t * net, const pnal_cfg_t * pnal_cfg);

/**
 * Get network parameters (IP address, netmask etc)
 *
 * For example:
 *
 * IP address        Represented by
 * 1.0.0.0           0x01000000 = 16777216
 * 0.0.0.1           0x00000001 = 1
 *
 * @param interface_name   In:    Ethernet interface name, for example eth0
 * @param p_ipaddr         Out:   IPv4 address
 * @param p_netmask        Out:   Netmask
 * @param p_gw             Out:   Default gateway
 * @param hostname         Out:   Host name, for example my_laptop_4. Existing
 *                                buffer should have size
 *                                PNAL_HOSTNAME_MAX_SIZE.
 * @return  0  if the operation succeeded.
 *          -1 if an error occurred.
 */
int pnal_get_ip_suite (
   const char * interface_name,
   pnal_ipaddr_t * p_ipaddr,
   pnal_ipaddr_t * p_netmask,
   pnal_ipaddr_t * p_gw,
   char * hostname);

/**
 * Read the IP address as an integer. For IPv4.
 *
 * For example:
 *
 * IP address        Represented by
 * 1.0.0.0           0x01000000 = 16777216
 * 0.0.0.1           0x00000001 = 1
 *
 * @param interface_name   In:    Name of network interface
 * @return IP address on success and
 *         0 if an error occurred
 */
pnal_ipaddr_t pnal_get_ip_address (const char * interface_name);

/**
 * Read the netmask as an integer. For IPv4.
 *
 * @param interface_name   In:    Name of network interface
 * @return netmask
 */
pnal_ipaddr_t pnal_get_netmask (const char * interface_name);

/**
 * Read the default gateway address as an integer. For IPv4.
 *
 * Assumes the default gateway is found on .1 on same subnet as the IP address.
 *
 * @param interface_name   In:    Name of network interface
 * @return netmask
 */
pnal_ipaddr_t pnal_get_gateway (const char * interface_name);

/**
 * Read the MAC address.
 *
 * @param interface_name   In:    Name of network interface
 * @param p_mac            Out:   MAC address
 *
 * @return 0 on success and
 *         -1 if no such interface is available
 */
int pnal_get_macaddress (const char * interface_name, pnal_ethaddr_t * p_mac);

/**
 * Read the current host name
 *
 * The resulting string is null terminated.
 *
 * @param hostname         Out:   Host name, for example my_laptop_4. Existing
 *                                buffer should have size
 *                                PNAL_HOSTNAME_MAX_SIZE.
 * @return 0 on success and
 *         -1 if an error occurred
 */
int pnal_get_hostname (char * hostname);

/**
 * Set network parameters (IP address, netmask etc)
 *
 * For example:
 *
 * IP address        Represented by
 * 1.0.0.0           0x01000000 = 16777216
 * 0.0.0.1           0x00000001 = 1
 *
 * @param interface_name   In:    Ethernet interface name, for example eth0
 * @param p_ipaddr         In:    IPv4 address
 * @param p_netmask        In:    Netmask
 * @param p_gw             In:    Default gateway
 * @param hostname         In:    Host name, for example my_laptop_4
 * @param permanent        In:    1 if changes are permanent, or 0 if temporary
 * @return  0  if the operation succeeded.
 *          -1 if an error occurred.
 */
int pnal_set_ip_suite (
   const char * interface_name,
   const pnal_ipaddr_t * p_ipaddr,
   const pnal_ipaddr_t * p_netmask,
   const pnal_ipaddr_t * p_gw,
   const char * hostname,
   bool permanent);

#ifdef __cplusplus
}
#endif

#endif /* PNAL_H */
